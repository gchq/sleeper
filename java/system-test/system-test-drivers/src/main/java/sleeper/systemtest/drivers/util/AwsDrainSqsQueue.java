/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package sleeper.systemtest.drivers.util;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.Message;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;
import java.util.stream.Collector;
import java.util.stream.IntStream;
import java.util.stream.LongStream;
import java.util.stream.Stream;

import static java.util.function.Predicate.not;
import static java.util.stream.Collectors.counting;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.summingLong;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toUnmodifiableList;

public class AwsDrainSqsQueue {
    public static final Logger LOGGER = LoggerFactory.getLogger(AwsDrainSqsQueue.class);
    private static final ExecutorService EXECUTOR = createThreadPool();

    private static ExecutorService createThreadPool() {
        ThreadPoolExecutor pool = new ThreadPoolExecutor(
                40, 40,
                60L, TimeUnit.SECONDS,
                new LinkedBlockingQueue<Runnable>());
        pool.allowCoreThreadTimeOut(true);
        return pool;
    }

    private final ReceiveMessages receiveMessages;
    private final int numThreads;
    private final int messagesPerBatchPerThread;
    private final int messagesPerReceive;
    private final int waitTimeSeconds;

    private AwsDrainSqsQueue(Builder builder) {
        receiveMessages = builder.receiveMessages;
        messagesPerBatchPerThread = builder.messagesPerBatchPerThread;
        messagesPerReceive = builder.messagesPerReceive;
        numThreads = builder.numThreads;
        waitTimeSeconds = builder.waitTimeSeconds;
    }

    public static AwsDrainSqsQueue forSystemTests(SqsClient sqsClient) {
        return withClient(sqsClient)
                .numThreads(10)
                .messagesPerBatchPerThread(10_000)
                .waitTimeSeconds(20) // Note that this is the maximum wait time
                .build();
    }

    public static AwsDrainSqsQueue forLocalStackTests(SqsClient sqsClient) {
        return withClient(sqsClient)
                .numThreads(2)
                .messagesPerBatchPerThread(10)
                .waitTimeSeconds(1)
                .build();
    }

    private static Builder withClient(SqsClient sqsClient) {
        return builder().sqsClient(sqsClient);
    }

    public static Builder builder() {
        return new Builder();
    }

    public Stream<Message> drain(String queueUrl) {
        LOGGER.info("Draining queue until empty: {}", queueUrl);
        return Stream.iterate(
                receiveMessageBatch(queueUrl), not(List::isEmpty), lastJobs -> receiveMessageBatch(queueUrl))
                .flatMap(List::stream);
    }

    public Stream<Message> drainExpectingMessages(int expectedMessages, String queueUrl) {
        LOGGER.info("Draining queue until empty, expecting {} messages: {}", queueUrl);
        return Stream.iterate(
                ReceiveBatchResult.first(receiveMessageBatch(queueUrl)),
                lastResult -> lastResult.hasMessages() || lastResult.totalMessages() < expectedMessages,
                lastResult -> lastResult.receiveNext(() -> receiveMessageBatch(queueUrl)))
                .flatMap(ReceiveBatchResult::stream);
    }

    public EmptyQueueResults empty(List<String> queueUrls) {
        LOGGER.info("Emptying queues: {}", queueUrls);
        EmptyQueueResults results = Stream.iterate(
                emptyMessageBatch(queueUrls),
                EmptyQueueResults::isAnyMessagesDeleted,
                lastResults -> emptyMessageBatch(lastResults.getQueueUrlsWithMessages()))
                .reduce(EmptyQueueResults.none(queueUrls),
                        (results1, results2) -> EmptyQueueResults.combine(Stream.of(results1, results2)));
        LOGGER.info("Deleted messages from queues: {}", results);
        return results;
    }

    public void empty(String queueUrl) {
        LOGGER.info("Emptying queue: {}", queueUrl);
        long count = LongStream.iterate(
                emptyMessageBatch(queueUrl), n -> n > 0, lastCount -> emptyMessageBatch(queueUrl))
                .sum();
        LOGGER.info("Deleted {} messages from queue: {}", count, queueUrl);
    }

    private List<Message> receiveMessageBatch(String queueUrl) {
        List<Message> messages = receiveOnThreads(queueUrl, toUnmodifiableList())
                .flatMap(List::stream).toList();
        LOGGER.info("Received a batch of {} messages from queue: {}", messages.size(), queueUrl);
        return messages;
    }

    private long emptyMessageBatch(String queueUrl) {
        return receiveOnThreads(queueUrl, counting()).mapToLong(n -> n).sum();
    }

    private EmptyQueueResults emptyMessageBatch(List<String> queueUrls) {
        List<List<String>> threadQueueBuckets = bucketQueuesByThreads(queueUrls);
        List<Callable<EmptyQueueResults>> callables = threadQueueBuckets.stream()
                .map(threadQueueUrls -> (Callable<EmptyQueueResults>) () -> emptyMessageBatchBucket(threadQueueUrls))
                .toList();
        return EmptyQueueResults.combine(streamInvokeAll(callables));
    }

    private List<List<String>> bucketQueuesByThreads(List<String> queueUrls) {
        int numBuckets = Math.min(numThreads, queueUrls.size());
        List<List<String>> threadBuckets = IntStream.range(0, numBuckets)
                .mapToObj(i -> (List<String>) new ArrayList<String>())
                .toList();
        Queue<String> remaining = new ArrayDeque<>(queueUrls);
        int thread = 0;
        for (String queueUrl = remaining.poll(); queueUrl != null; queueUrl = remaining.poll()) {
            threadBuckets.get(thread).add(queueUrl);
            thread = (thread + 1) % numBuckets;
        }
        return threadBuckets;
    }

    private EmptyQueueResults emptyMessageBatchBucket(List<String> queueUrls) {
        return EmptyQueueResults.fromOneThreadBatch(queueUrls.stream()
                .map(queueUrl -> new EmptyQueueResult(queueUrl, receiveMessageBatchOneThread(queueUrl, counting()))));
    }

    private <A, R> Stream<R> receiveOnThreads(String queueUrl, Collector<Message, A, R> threadCollector) {
        return streamInvokeAll(IntStream.range(0, numThreads)
                .mapToObj(i -> (Callable<R>) () -> receiveMessageBatchOneThread(queueUrl, threadCollector))
                .toList());
    }

    private <R> Stream<R> streamInvokeAll(List<Callable<R>> callables) {
        try {
            List<Future<R>> results = EXECUTOR.invokeAll(callables);
            return results.stream()
                    .map(future -> {
                        try {
                            return future.get();
                        } catch (InterruptedException e) {
                            Thread.currentThread().interrupt();
                            throw new RuntimeException(e);
                        } catch (ExecutionException e) {
                            throw new RuntimeException(e);
                        }
                    });
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException(e);
        }
    }

    private <A, R> R receiveMessageBatchOneThread(String queueUrl, Collector<Message, A, R> collector) {
        return Stream.iterate(
                ReceiveBatchResult.first(receiveMessages(queueUrl)),
                lastResult -> lastResult.hasMessages(),
                lastResult -> lastResult.totalMessages() < messagesPerBatchPerThread
                        ? lastResult.receiveNext(() -> receiveMessages(queueUrl))
                        : lastResult.noMessagesNext())
                .flatMap(ReceiveBatchResult::stream)
                .collect(collector);
    }

    private List<Message> receiveMessages(String queueUrl) {
        return receiveMessages.receiveAndDeleteMessages(queueUrl, messagesPerReceive, waitTimeSeconds);
    }

    private record ReceiveBatchResult(List<Message> messages, int totalMessages, int totalReceives) {

        static ReceiveBatchResult first(List<Message> messages) {
            return new ReceiveBatchResult(messages, messages.size(), 1);
        }

        boolean hasMessages() {
            return !messages.isEmpty();
        }

        ReceiveBatchResult receiveNext(Supplier<List<Message>> receiveMessages) {
            List<Message> receivedMessages = receiveMessages.get();
            int newTotal = totalMessages + receivedMessages.size();
            LOGGER.info("Recevied {} messages, total of {} for batch", receivedMessages.size(), newTotal);
            return new ReceiveBatchResult(receivedMessages, newTotal, totalReceives + 1);
        }

        ReceiveBatchResult noMessagesNext() {
            return new ReceiveBatchResult(List.of(), totalMessages, totalReceives);
        }

        Stream<Message> stream() {
            return messages.stream();
        }
    }

    public record EmptyQueueResult(String queueUrl, long messages) {
    }

    public static class EmptyQueueResults {
        private final Map<String, Long> messagesByQueueUrl;

        private EmptyQueueResults(Map<String, Long> messagesByQueueUrl) {
            this.messagesByQueueUrl = messagesByQueueUrl;
        }

        private static EmptyQueueResults none(List<String> queueUrls) {
            return new EmptyQueueResults(queueUrls.stream()
                    .collect(toMap(queueUrl -> queueUrl, queueUrl -> 0L)));
        }

        private static EmptyQueueResults fromOneThreadBatch(Stream<EmptyQueueResult> results) {
            return new EmptyQueueResults(
                    results.collect(toMap(EmptyQueueResult::queueUrl, EmptyQueueResult::messages)));
        }

        private static EmptyQueueResults combine(Stream<EmptyQueueResults> results) {
            return new EmptyQueueResults(
                    results.flatMap(EmptyQueueResults::stream)
                            .collect(groupingBy(EmptyQueueResult::queueUrl,
                                    summingLong(EmptyQueueResult::messages))));
        }

        public boolean isAnyMessagesDeleted() {
            return messagesByQueueUrl.values().stream().anyMatch(count -> count > 0);
        }

        public List<String> getQueueUrlsWithMessages() {
            return stream()
                    .filter(result -> result.messages() > 0)
                    .map(EmptyQueueResult::queueUrl)
                    .toList();
        }

        public Stream<EmptyQueueResult> stream() {
            return messagesByQueueUrl.entrySet().stream()
                    .map(entry -> new EmptyQueueResult(entry.getKey(), entry.getValue()));
        }

        public long getMessagesDeleted(String queueUrl) {
            return messagesByQueueUrl.getOrDefault(queueUrl, 0L);
        }

        @Override
        public String toString() {
            return messagesByQueueUrl.toString();
        }
    }

    public static class Builder {

        private ReceiveMessages receiveMessages;
        private int numThreads = 1;
        private int messagesPerBatchPerThread = 10_000;
        private int messagesPerReceive = 10;
        private int waitTimeSeconds = 10;

        private Builder() {
        }

        public Builder sqsClient(SqsClient sqsClient) {
            return receiveMessages(ReceiveMessages.from(sqsClient));
        }

        public Builder receiveMessages(ReceiveMessages receiveMessages) {
            this.receiveMessages = receiveMessages;
            return this;
        }

        public Builder numThreads(int numThreads) {
            this.numThreads = numThreads;
            return this;
        }

        public Builder messagesPerBatchPerThread(int messagesPerBatchPerThread) {
            this.messagesPerBatchPerThread = messagesPerBatchPerThread;
            return this;
        }

        public Builder messagesPerReceive(int messagesPerReceive) {
            this.messagesPerReceive = messagesPerReceive;
            return this;
        }

        public Builder waitTimeSeconds(int waitTimeSeconds) {
            this.waitTimeSeconds = waitTimeSeconds;
            return this;
        }

        public AwsDrainSqsQueue build() {
            return new AwsDrainSqsQueue(this);
        }
    }
}
