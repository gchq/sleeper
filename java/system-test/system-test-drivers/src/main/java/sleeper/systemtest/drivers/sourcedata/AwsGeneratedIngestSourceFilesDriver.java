/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.drivers.sourcedata;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import software.amazon.awssdk.services.s3.S3Client;
import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
import software.amazon.awssdk.services.s3.model.S3Object;

import sleeper.configuration.properties.S3InstanceProperties;
import sleeper.systemtest.dsl.instance.DeployedSystemTestResources;
import sleeper.systemtest.dsl.sourcedata.GeneratedIngestSourceFiles;
import sleeper.systemtest.dsl.sourcedata.GeneratedIngestSourceFilesDriver;

import java.util.List;
import java.util.stream.Collectors;

import static java.util.function.Predicate.not;

public class AwsGeneratedIngestSourceFilesDriver implements GeneratedIngestSourceFilesDriver {
    private static final Logger LOGGER = LoggerFactory.getLogger(AwsGeneratedIngestSourceFilesDriver.class);

    private final DeployedSystemTestResources context;
    private final S3Client s3Client;

    public AwsGeneratedIngestSourceFilesDriver(DeployedSystemTestResources context, S3Client s3Client) {
        this.context = context;
        this.s3Client = s3Client;
    }

    public GeneratedIngestSourceFiles findGeneratedFiles() {
        String bucketName = context.getSystemTestBucketName();
        List<String> objectKeys = s3Client.listObjectsV2Paginator(builder -> builder.bucket(bucketName).prefix("ingest/"))
                .contents().stream()
                .map(S3Object::key)
                .collect(Collectors.toUnmodifiableList());
        LOGGER.info("Found ingest objects in source bucket: {}", objectKeys.size());
        return new GeneratedIngestSourceFiles(bucketName, objectKeys);
    }

    public void emptyBucket() {
        String bucketName = context.getSystemTestBucketName();
        List<ObjectIdentifier> objects = s3Client.listObjectsV2Paginator(builder -> builder.bucket(bucketName))
                .contents().stream().map(S3Object::key)
                .filter(not(S3InstanceProperties.S3_INSTANCE_PROPERTIES_FILE::equals))
                .map(key -> ObjectIdentifier.builder().key(key).build())
                .collect(Collectors.toList());
        if (!objects.isEmpty()) {
            s3Client.deleteObjects(builder -> builder.bucket(bucketName)
                    .delete(deleteBuilder -> deleteBuilder.objects(objects)));
        }
    }
}
