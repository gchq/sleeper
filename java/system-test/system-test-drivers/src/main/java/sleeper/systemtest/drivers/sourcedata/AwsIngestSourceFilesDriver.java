/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.drivers.sourcedata;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.hadoop.ParquetWriter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.record.Row;
import sleeper.core.schema.Schema;
import sleeper.parquet.record.ParquetRecordWriterFactory;
import sleeper.sketches.Sketches;
import sleeper.sketches.store.S3SketchesStore;
import sleeper.systemtest.drivers.util.SystemTestClients;
import sleeper.systemtest.dsl.sourcedata.IngestSourceFilesDriver;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.Iterator;

public class AwsIngestSourceFilesDriver implements IngestSourceFilesDriver {
    private static final Logger LOGGER = LoggerFactory.getLogger(AwsIngestSourceFilesDriver.class);
    private final SystemTestClients clients;

    public AwsIngestSourceFilesDriver(SystemTestClients clients) {
        this.clients = clients;
    }

    public void writeFile(
            InstanceProperties instanceProperties, TableProperties tableProperties,
            String path, boolean writeSketches, Iterator<Row> records) {
        Schema schema = tableProperties.getSchema();
        Configuration conf = clients.createHadoopConf(instanceProperties, tableProperties);
        Sketches sketches = Sketches.from(schema);
        LOGGER.info("Writing to {}", path);
        try (ParquetWriter<Row> writer = ParquetRecordWriterFactory.createParquetRecordWriter(
                new Path(path), tableProperties, conf)) {
            for (Row record : (Iterable<Row>) () -> records) {
                sketches.update(record);
                writer.write(record);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        if (writeSketches) {
            LOGGER.info("Writing sketches");
            S3SketchesStore.createWriteOnly(clients.getS3TransferManager())
                    .saveFileSketches(path, schema, sketches);
        }
    }
}
