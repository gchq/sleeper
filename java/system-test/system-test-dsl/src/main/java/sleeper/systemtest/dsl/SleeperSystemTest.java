/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.dsl;

import sleeper.core.properties.instance.InstanceProperties;
import sleeper.core.properties.table.TableProperties;
import sleeper.core.properties.table.TableProperty;
import sleeper.core.properties.validation.OptionalStack;
import sleeper.core.record.Record;
import sleeper.core.schema.Schema;
import sleeper.systemtest.dsl.compaction.SystemTestCompaction;
import sleeper.systemtest.dsl.gc.SystemTestGarbageCollection;
import sleeper.systemtest.dsl.ingest.SystemTestIngest;
import sleeper.systemtest.dsl.instance.SystemTestInstanceConfiguration;
import sleeper.systemtest.dsl.instance.SystemTestOptionalStacks;
import sleeper.systemtest.dsl.instance.SystemTestParameters;
import sleeper.systemtest.dsl.instance.SystemTestTableFiles;
import sleeper.systemtest.dsl.instance.SystemTestTables;
import sleeper.systemtest.dsl.metrics.SystemTestTableMetrics;
import sleeper.systemtest.dsl.partitioning.SystemTestPartitioning;
import sleeper.systemtest.dsl.python.SystemTestPythonApi;
import sleeper.systemtest.dsl.query.SystemTestQuery;
import sleeper.systemtest.dsl.reporting.SystemTestReporting;
import sleeper.systemtest.dsl.sourcedata.GenerateNumberedRecords;
import sleeper.systemtest.dsl.sourcedata.GenerateNumberedValueOverrides;
import sleeper.systemtest.dsl.sourcedata.RecordNumbers;
import sleeper.systemtest.dsl.sourcedata.SystemTestCluster;
import sleeper.systemtest.dsl.sourcedata.SystemTestLocalFiles;
import sleeper.systemtest.dsl.sourcedata.SystemTestSourceFiles;
import sleeper.systemtest.dsl.statestore.SystemTestStateStore;

import java.nio.file.Path;
import java.util.Collection;
import java.util.Map;
import java.util.stream.LongStream;

import static sleeper.core.properties.table.TableProperty.TABLE_ONLINE;

/**
 * The entry point that all system tests use to interact with the system. This is the starting point for all steps of
 * the Domain Specific Language (DSL) for system tests. The purpose of this is to make it as easy as possible to read
 * and write system tests. It should make it easy to find and reuse steps to interact with the systems you care about
 * for your tests.
 * <p>
 * It does this by delegating to drivers to actually interact with the system. The DSL only defines
 * the steps and language to be used by tests, and not any of the implemented behaviour of the steps.
 * <p>
 * This class should match as closely as possible with a diagram of the components of the system. You can
 * expect to find the system you care about through a method on this class. Some core features of the system are
 * on this class directly, but we try to limit its size by grouping methods into components which you access through
 * a method, eg. {@link #ingest()}.
 * <p>
 * Most tests will use steps from different systems, but where multiple steps run against the same system in succession
 * we can use method chaining to avoid accessing that system repeatedly. If you go from one system to another and back,
 * assume you should re-access the first system from here again the second time.
 * Try to avoid assigning variables except for data you want to reuse.
 */
public class SleeperSystemTest {

    private final SystemTestParameters parameters;
    private final SystemTestDrivers baseDrivers;
    private final SystemTestContext context;

    public SleeperSystemTest(SystemTestParameters parameters, SystemTestDrivers baseDrivers, SystemTestContext context) {
        this.parameters = parameters;
        this.baseDrivers = baseDrivers;
        this.context = context;
    }

    public void connectToInstanceAddOnlineTable(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
        context.instance().addDefaultTables(Map.of(TABLE_ONLINE, "true"));
    }

    public void connectToInstanceAddOfflineTable(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
        context.instance().addDefaultTables(Map.of(TABLE_ONLINE, "false"));
    }

    public void connectToInstanceNoTables(SystemTestInstanceConfiguration configuration) {
        context.instance().connectTo(configuration);
    }

    public InstanceProperties instanceProperties() {
        return context.instance().getInstanceProperties();
    }

    public TableProperties tableProperties() {
        return context.instance().getTableProperties();
    }

    public void updateTableProperties(Map<TableProperty, String> values) {
        context.instance().updateTableProperties(values);
    }

    public SystemTestSourceFiles sourceFiles() {
        return new SystemTestSourceFiles(context.instance(), context.sourceFiles(), baseDrivers.sourceFiles(context));
    }

    public SystemTestTableFiles tableFiles() {
        return new SystemTestTableFiles(context.instance());
    }

    public SystemTestPartitioning partitioning() {
        return new SystemTestPartitioning(context);
    }

    public SystemTestIngest ingest() {
        return new SystemTestIngest(context, baseDrivers);
    }

    public SystemTestQuery query() {
        return new SystemTestQuery(context, baseDrivers);
    }

    public SystemTestQuery directQuery() {
        return query().direct();
    }

    public SystemTestCompaction compaction() {
        return new SystemTestCompaction(context, baseDrivers);
    }

    public SystemTestGarbageCollection garbageCollection() {
        return new SystemTestGarbageCollection(context);
    }

    public SystemTestReporting reporting() {
        return new SystemTestReporting(context);
    }

    public SystemTestTableMetrics tableMetrics() {
        return new SystemTestTableMetrics(context.instance().adminDrivers().tableMetrics(context));
    }

    public SystemTestCluster systemTestCluster() {
        return new SystemTestCluster(context, baseDrivers);
    }

    public SystemTestPythonApi pythonApi() {
        return new SystemTestPythonApi(context);
    }

    public SystemTestLocalFiles localFiles(Path tempDir) {
        return new SystemTestLocalFiles(context.instance(), baseDrivers.localFiles(context), tempDir);
    }

    public void setGeneratorOverrides(GenerateNumberedValueOverrides overrides) {
        context.instance().setGeneratorOverrides(overrides);
    }

    public GenerateNumberedRecords numberedRecords() {
        return context.instance().numberedRecords();
    }

    public Iterable<Record> generateNumberedRecords(LongStream numbers) {
        return context.instance().numberedRecords().iterableFrom(numbers);
    }

    public Iterable<Record> generateNumberedRecords(Schema schema, LongStream numbers) {
        return context.instance().numberedRecords(schema).iterableFrom(numbers);
    }

    public RecordNumbers scrambleNumberedRecords(LongStream longStream) {
        return RecordNumbers.scrambleNumberedRecords(longStream);
    }

    public Path getSplitPointsDirectory() {
        return parameters.getScriptsDirectory()
                .resolve("test/splitpoints");
    }

    public void enableOptionalStack(OptionalStack stack) {
        new SystemTestOptionalStacks(context.instance()).addOptionalStack(stack);
    }

    public void enableOptionalStacks(Collection<OptionalStack> stacks) {
        new SystemTestOptionalStacks(context.instance()).addOptionalStacks(stacks);
    }

    public void disableOptionalStack(OptionalStack stack) {
        new SystemTestOptionalStacks(context.instance()).removeOptionalStack(stack);
    }

    public void disableOptionalStacks(Collection<OptionalStack> stacks) {
        new SystemTestOptionalStacks(context.instance()).removeOptionalStacks(stacks);
    }

    public SystemTestStateStore stateStore() {
        return new SystemTestStateStore(context);
    }

    public SystemTestTables tables() {
        return new SystemTestTables(context.instance());
    }

    public SleeperSystemTest table(String name) {
        context.instance().setCurrentTable(name);
        return this;
    }
}
