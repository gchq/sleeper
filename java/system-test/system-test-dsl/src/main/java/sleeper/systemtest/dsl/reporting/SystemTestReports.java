/*
 * Copyright 2022-2025 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.dsl.reporting;

import sleeper.systemtest.dsl.SystemTestContext;
import sleeper.systemtest.dsl.SystemTestDrivers;
import sleeper.systemtest.dsl.util.TestContext;

import java.util.ArrayList;
import java.util.List;

public class SystemTestReports {

    private final ReportingContext context;
    private final List<SystemTestReport> reports;

    private SystemTestReports(Builder builder) {
        context = builder.context;
        reports = builder.reports;
    }

    public static Builder builder(ReportingContext context) {
        return new Builder(context);
    }

    public static SystemTestBuilder builder(SystemTestContext context) {
        return new SystemTestBuilder(context);
    }

    public void print(TestContext testContext) {
        context.print(testContext, (out, startTime) -> reports.forEach(report -> report.print(out, startTime)));
    }

    public static class Builder {
        private final ReportingContext context;
        private final List<SystemTestReport> reports = new ArrayList<>();

        private Builder(ReportingContext context) {
            this.context = context;
        }

        public Builder report(SystemTestReport report) {
            reports.add(report);
            return this;
        }

        public SystemTestReports build() {
            return new SystemTestReports(this);
        }
    }

    public static class SystemTestBuilder extends Builder {

        private final PartitionReportDriver partitionDriver;
        private final IngestReportsDriver ingestDriver;
        private final CompactionReportsDriver compactionDriver;

        private SystemTestBuilder(SystemTestContext context) {
            super(context.reporting());
            SystemTestDrivers drivers = context.instance().adminDrivers();
            this.partitionDriver = drivers.partitionReports(context);
            this.ingestDriver = drivers.ingestReports(context);
            this.compactionDriver = drivers.compactionReports(context);
        }

        public Builder ingestTasksAndJobs() {
            return report(ingestDriver.tasksAndJobsReport());
        }

        public Builder ingestJobs() {
            return report(ingestDriver.jobsReport());
        }

        public Builder compactionTasksAndJobs() {
            return report(compactionDriver.tasksAndJobsReport());
        }

        public Builder partitionStatus() {
            return report(partitionDriver.statusReport());
        }
    }
}
