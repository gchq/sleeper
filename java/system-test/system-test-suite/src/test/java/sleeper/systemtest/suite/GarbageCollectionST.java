/*
 * Copyright 2022-2024 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sleeper.systemtest.suite;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import sleeper.compaction.strategy.impl.BasicCompactionStrategy;
import sleeper.systemtest.dsl.SleeperSystemTest;
import sleeper.systemtest.dsl.extension.AfterTestPurgeQueues;
import sleeper.systemtest.dsl.extension.AfterTestReports;
import sleeper.systemtest.dsl.reporting.SystemTestReports;
import sleeper.systemtest.dsl.sourcedata.RecordNumbers;
import sleeper.systemtest.suite.testutil.SystemTest;

import java.nio.file.Path;
import java.util.Map;
import java.util.stream.LongStream;

import static org.assertj.core.api.Assertions.assertThat;
import static sleeper.configuration.properties.instance.CdkDefinedInstanceProperty.COMPACTION_JOB_QUEUE_URL;
import static sleeper.configuration.properties.table.TableProperty.COMPACTION_FILES_BATCH_SIZE;
import static sleeper.configuration.properties.table.TableProperty.COMPACTION_STRATEGY_CLASS;
import static sleeper.configuration.properties.table.TableProperty.GARBAGE_COLLECTOR_DELAY_BEFORE_DELETION;
import static sleeper.core.testutils.printers.FileReferencePrinter.printFiles;
import static sleeper.systemtest.suite.fixtures.SystemTestInstance.MAIN;

@SystemTest
public class GarbageCollectionST {
    @TempDir
    private Path tempDir;

    @BeforeEach
    void setUp(SleeperSystemTest sleeper, AfterTestReports reporting, AfterTestPurgeQueues purgeQueues) {
        sleeper.connectToInstance(MAIN);
        reporting.reportIfTestFailed(SystemTestReports.SystemTestBuilder::compactionTasksAndJobs);
        purgeQueues.purgeIfTestFailed(COMPACTION_JOB_QUEUE_URL);
    }

    @Test
    void shouldGarbageCollectFilesAfterCompaction(SleeperSystemTest sleeper) {
        // Given
        sleeper.updateTableProperties(Map.of(
                COMPACTION_STRATEGY_CLASS, BasicCompactionStrategy.class.getName(),
                COMPACTION_FILES_BATCH_SIZE, "5",
                GARBAGE_COLLECTOR_DELAY_BEFORE_DELETION, "0"));
        RecordNumbers numbers = sleeper.scrambleNumberedRecords(LongStream.range(0, 50));
        sleeper.ingest().direct(tempDir)
                .numberedRecords(numbers.range(0, 10))
                .numberedRecords(numbers.range(10, 20))
                .numberedRecords(numbers.range(20, 30))
                .numberedRecords(numbers.range(30, 40))
                .numberedRecords(numbers.range(40, 50));
        sleeper.compaction().createJobs(1).invokeTasks(1).waitForJobs();

        // When
        sleeper.garbageCollection().invoke().waitFor();

        // Then
        assertThat(sleeper.directQuery().allRecordsInTable())
                .containsExactlyInAnyOrderElementsOf(sleeper.generateNumberedRecords(LongStream.range(0, 50)));
        Approvals.verify(printFiles(sleeper.partitioning().tree(), sleeper.tableFiles().all()));
    }
}
