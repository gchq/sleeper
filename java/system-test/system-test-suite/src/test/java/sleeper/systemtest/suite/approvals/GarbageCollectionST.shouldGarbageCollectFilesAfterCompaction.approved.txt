Unreferenced files: 0
Referenced files: 1
File references: 1
Partition at root: 50 records in file 1
